cloud/openstack/os_aggregate.py

```diff
@@ -0,0 +1,196 @@
+options:
+  name:
+    description: Name of the aggregate.
+    required: true
+  metadata:
+    description: Metadata dict.
+    required: false
+    default: None
+  availability_zone:
+    description: Availability zone to create aggregate into.
+    required: false
+    default: None
+  hosts:
+    description: List of hosts to set for an aggregate.
+    required: false
+    default: None
+  state:
+    description: Should the resource be present or absent.
+    choices: [present, absent]
+    default: present
+requirements:
+    - "python >= 2.6"
+    - "shade"
+'''
+
+EXAMPLES = '''
+# Create a host aggregate
+- os_aggregate:
+    cloud: mycloud
+    state: present
+    name: db_aggregate
+    hosts:
+      - host1
+      - host2
+    metadata:
+      type: dbcluster
+# Delete an aggregate
+- os_aggregate:
+    cloud: mycloud
+    state: absent
+    name: db_aggregate
+'''
+
+RETURN = '''
+
+'''
+
+def _needs_update(module, aggregate):
+    new_metadata = (module.params['metadata'] or {})
+    new_metadata['availability_zone'] = module.params['availability_zone']
+
+    if (module.params['name'] != aggregate.name) or \
+        (module.params['hosts'] is not None and module.params['hosts'] != aggregate.hosts) or \
+        (module.params['availability_zone'] is not None and module.params['availability_zone'] != aggregate.availability_zone) or \
+        (module.params['metadata'] is not None and new_metadata != aggregate.metadata):
+        return True
+
+    return False
+
+def _system_state_change(module, aggregate):
+    state = module.params['state']
+    if state == 'absent' and aggregate:
+        return True
+
+    if state == 'present':
+        if aggregate is None:
+            return True
+        return _needs_update(module, aggregate)
+
+    return False
+
+def main():
+    argument_spec = openstack_full_argument_spec(
+        name=dict(required=True),
+        metadata=dict(required=False, default=None, type='dict'),
+        availability_zone=dict(required=False, default=None),
+        hosts=dict(required=False, default=None, type='list'),
+        state=dict(default='present', choices=['absent', 'present']),
+    )
+
+    module_kwargs = openstack_module_kwargs()
+    module = AnsibleModule(argument_spec,
+                           supports_check_mode=True,
+                           **module_kwargs)
+
+    if not HAS_SHADE:
+        module.fail_json(msg='shade is required for this module')
+    if StrictVersion(shade.__version__) < StrictVersion('1.6.0'):
+        module.fail_json(msg="To utilize this module, the installed version of"
+                             "the shade library MUST be >=1.6.0")
+
+    name = module.params['name']
+    metadata = module.params['metadata']
+    availability_zone = module.params['availability_zone']
+    hosts = module.params['hosts']
+    state = module.params['state']
+
+    if metadata is not None:
+        metadata.pop('availability_zone', None)
```

Comment:
Does putting availability_zone in the metadata conflict with the ansible module parameter then? If so, should we toss an error if the user tries to supply one? (For that matter, should we update shade's set_aggregate_metadata to validate and throw an error?)